### PR TITLE
Exceptions to repeating events

### PR DESCRIPTION
**Pull request by nokhub**

### CODE REVIEW

File: src/api/worker/offline/migrations/sys-v84.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { OfflineMigration } from "../OfflineStorageMigrator.js"
+import { OfflineStorage } from "../OfflineStorage.js"
+import { migrateAllElements } from "../StandardMigrations.js"
+import { createMissedNotification, MissedNotificationTypeRef } from "../../../entities/sys/TypeRefs.js"
+
+export const sys84: OfflineMigration = {
+	app: "sys",
+	version: 84,
+	async migrate(storage: OfflineStorage) {
+		await migrateAllElements(MissedNotificationTypeRef, storage, [createMissedNotification])
```
Comment (charlag) [inline src/api/worker/offline/migrations/sys-v84.ts:10]: missed notifications are not cached (they don't even really exist haha)

Comment (ganthern) [reply src/api/worker/offline/migrations/sys-v84.ts]: we put a no-op migration with an explanation because the compile time check does not know about this

File: src/calendar/view/CalendarEventPopup.ts [hunk 2]
```diff
@@ -150,26 +143,62 @@ export class CalendarEventPopup implements ModalComponent {
...
+			attachDropdown({
+				mainButtonAttrs: {
+					title: "delete_action",
+					icon: Icons.Trash,
+					colors: ButtonColor.DrawerNav,
+				},
+				childAttrs: () => [
+					{
+						// FIXME add to phrase
+						label: "deleteSingleEventRecurrence_msg",
```
Comment (charlag) [inline src/calendar/view/CalendarEventPopup.ts:180]: FIXME

File: src/calendar/view/CalendarEventPopup.ts [hunk 3]
```diff
@@ -150,26 +143,62 @@ export class CalendarEventPopup implements ModalComponent {
...
+						click: () => this.addExclusion(),
+					},
+					{
+						// FIXME add to phrase
+						label: "deleteAllEventRecurrence_msg",
```
Comment (charlag) [inline src/calendar/view/CalendarEventPopup.ts:185]: FIXME

File: src/calendar/view/CalendarViewModel.ts [hunk 4]
```diff
@@ -58,7 +58,9 @@ export type EventsOnDays = {
 	longEvents: Array<CalendarEvent>
 }
 export type DraggedEvent = {
+	// the event instance the user grabbed with the mouse
```
Comment (charlag) [inline src/calendar/view/CalendarViewModel.ts:61]: convert them to jsdoc

File: test/tests/calendar/CalendarEventViewModelTest.ts [hunk 5]
```diff
@@ -3053,6 +3061,58 @@ o.spec("CalendarEventViewModel", function () {
...
+					}),
+				}),
+			})
+
+			viewModel.deleteExcludedDates()
+			o(viewModel.repeat?.excludedDates).deepEquals([])
+		})
+	})
+	o.spec("excludeOccurences", function () {
+		// TODO
```
Comment (charlag) [inline test/tests/calendar/CalendarEventViewModelTest.ts:3082]: TODO

File: test/tests/calendar/CalendarImporterTest.ts [hunk 6]
```diff
@@ -1138,4 +1142,39 @@ END:VCALENDAR`
...
+				serializeExcludedDates(
+					[createDateWrapper({ date: new Date("2023-01-14T22:00:00Z") }), createDateWrapper({ date: new Date("2023-01-21T22:00:00Z") })],
+					"Europe/Berlin",
+				),
+			).deepEquals(["EXDATE;TZID=Europe/Berlin:20230114T230000,20230121T230000"])
+		})
+	})
 })
+
+// const parsedDates = parseExDates([
```
Comment (charlag) [inline test/tests/calendar/CalendarImporterTest.ts:1168]: commented out

File: test/tests/calendar/CalendarUtilsTest.ts [hunk 7]
```diff
@@ -789,6 +792,154 @@ o.spec("calendar utils tests", function () {
 			).equals(CalendarEventValidity.Valid)
 		})
 	})
+
+	o.spec("exclusions", function () {
```
Comment (charlag) [inline test/tests/calendar/CalendarUtilsTest.ts:798]: should the spec be `calendarEventHasMoreThanOneOccurrencesLeft` ?

File: src/calendar/date/CalendarUtils.ts [hunk 8]
```diff
@@ -950,18 +951,19 @@ export function findNextAlarmOccurrence(
 
 	while (endType !== EndType.Count || occurrenceNumber < endValue) {
 		const occurrenceDate = incrementByRepeatPeriod(calcEventStart, frequency, interval * occurrenceNumber, isAllDayEvent ? localTimeZone : timeZone)
-
 		if (endDate && occurrenceDate.getTime() >= endDate.getTime()) {
 			return null
 		}
 
-		const alarmTime = calculateAlarmTime(occurrenceDate, alarmTrigger, localTimeZone)
+		if (exclusions.find((d) => d.getTime() === occurrenceDate.getTime()) == null) {
```
Comment (charlag) [inline src/calendar/date/CalendarUtils.ts:958]: incorrect semantics for `find` again

File: app-android/app/src/test/java/de/tutao/tutanota/AlarmModelTest.kt [hunk 9]
```diff
@@ -19,7 +19,7 @@ class AlarmModelTest {
 		val now = getDate(timeZone, 2019, 4, 2, 0, 0)
 		val eventStart = getDate(timeZone, 2019, 4, 2, 12, 0)
 		iterateAlarmOccurrences(now, timeZone, eventStart, eventStart, RepeatPeriod.WEEKLY,
-				1, EndType.NEVER, 0, AlarmTrigger.ONE_HOUR, timeZone
+				1, EndType.NEVER, 0, AlarmTrigger.ONE_HOUR, timeZone, emptyList()
```
Comment (charlag) [inline app-android/app/src/test/java/de/tutao/tutanota/AlarmModelTest.kt:22]: do we need tests with exclusions?

File: src/gui/TimePicker.ts [hunk 10]
```diff
@@ -31,7 +31,7 @@ export class TimePicker implements Component<TimePickerAttrs> {
 				times.push(timeStringFromParts(hour, minute, attrs.amPmFormat))
 			}
 		}
-
+		this._oldValue = attrs.time?.toString(false) ?? "--"
```
Comment (charlag) [inline src/gui/TimePicker.ts:34]: why is this necessary?

File: app-ios/tutanota/Sources/Alarms/AlarmModel.swift [hunk 11]
```diff
@@ -176,6 +177,13 @@ struct LazyEventSequence : Sequence, IteratorProtocol {
         occurenceDate: occurrenceDate
       )
       ocurrenceNumber += 1
+      
+      while exclusionNumber < repeatRule.excludedDates.count && repeatRule.excludedDates[exclusionNumber] < occurrenceDate {
+        exclusionNumber += 1
+      }
+      if (exclusionNumber < repeatRule.excludedDates.count && repeatRule.excludedDates[exclusionNumber] == occurrenceDate) {
+        return self.next()
```
Comment (charlag) [inline app-ios/tutanota/Sources/Alarms/AlarmModel.swift:185]: I hope this recursion won't bite us in the butt but I also see no other easy way

File: app-ios/tutanota/Sources/Alarms/RepeatRule.swift [hunk 12]
```diff
@@ -31,19 +31,45 @@ enum RepeatEndCondition : Equatable {
...
+  
+  public init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+    if let excludedDates = try container.decodeIfPresent([EncryptedDateWrapper].self, forKey: .excludedDates) {
+      self.excludedDates = excludedDates
+    } else {
+      self.excludedDates = [EncryptedDateWrapper]()
+    }
+    
+    self.frequency = try! container.decode(Base64.self, forKey: .frequency)
```
Comment (charlag) [inline app-ios/tutanota/Sources/Alarms/RepeatRule.swift:54]: why are those hard unwrapped?

File: app-ios/tutanota/Sources/Alarms/RepeatRule.swift [hunk 13]
```diff
@@ -56,5 +82,10 @@ extension RepeatRule {
       return try decrypt(base64: endValue, key: sessionKey)
     }
     self.endCondition = RepeatEndCondition(endType: endType, endValue: endValue ?? 0)
+    let decryptedExclusions: [Date] = try encrypted.excludedDates.map { encWrapper in
+      let timestamp: Int64 = try decrypt(base64: encWrapper.date, key: sessionKey)
```
Comment (charlag) [inline app-ios/tutanota/Sources/Alarms/RepeatRule.swift:86]: don't we have a function to decrypt a date?

File: app-ios/tutanotaTests/AlarmManagerTest.swift [hunk 14]
```diff
@@ -120,7 +120,7 @@ class AlarmManagerTest : XCTestCase {
         frequency: .daily,
         interval: 1,
         timeZone: dateProvider.timeZone.identifier,
-        endCondition: .count(times: 2)
+        endCondition: .count(times: 2), excludedDates: []
```
Comment (charlag) [inline app-ios/tutanotaTests/AlarmManagerTest.swift:123]: formatting

File: app-ios/tutanota/Sources/Alarms/RepeatRule.swift (hunk 12 above)
Comment (ganthern) [reply app-ios/tutanota/Sources/Alarms/RepeatRule.swift:54]: there's no reasonable default if they're missing

File: src/gui/TimePicker.ts (hunk 10 above)
Comment (ganthern) [reply src/gui/TimePicker.ts:34]: having a not-null assertion on a field that's not assigned in the constructor is no bueno, changing all the call sites to decide what the default should be if they don't have an oldValue is also not great. 

just having the oldValue be something that can't be a time is easiest for the native time picker to not immediately call _onSelected on every first render.

File: app-ios/tutanota/Sources/Alarms/RepeatRule.swift (hunk 12 above)
Comment (charlag) [reply app-ios/tutanota/Sources/Alarms/RepeatRule.swift:54]: I've meant that it uses `try!` instead of `try`

File: src/calendar/date/CalendarEventViewModel.ts [hunk 15]
```diff
@@ -492,12 +499,26 @@ export class CalendarEventViewModel {
...
 	}
 
+	setAllDay(newAllDay: boolean): void {
+		if (newAllDay === this._allDay) return
+		this._allDay = newAllDay
+		if (this.repeat == null) return
+		if (newAllDay) {
+			this.repeat.excludedDates = this.repeat.excludedDates.map((date) => getAllDayDateUTC(date))
+		} else {
+			const startTime = this.startTime ?? new Time(0, 0)
```
[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/date/CalendarEventViewModel.ts:517]: changed it to preemptively deleting the exclusions because we can't save anyway without losing them

File: src/calendar/date/CalendarEventViewModel.ts [hunk 16]
```diff
@@ -787,6 +810,43 @@ export class CalendarEventViewModel {
...
+	 * fire alarms.
+	 *
+	 * the list of exclusions is maintained sorted from earliest to latest.
+	 */
+	async excludeOccurrence(): Promise<void> {
+		const existingEvent = this.existingEvent
+		if (existingEvent == null) return
+		const selectedCalendar = this.selectedCalendar()
+		if (!selectedCalendar) return
+		const originalEvent = existingEvent.repeatRule ? await this._entityClient.load(CalendarEventTypeRef, existingEvent._id) : existingEvent
```
[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/date/CalendarEventViewModel.ts:828]: added a comment

File: app-ios/tutanota/Sources/Alarms/RepeatRule.swift (hunk 12 above)
Comment (nokhub) [reply app-ios/tutanota/Sources/Alarms/RepeatRule.swift:54]: right, fixed

File: src/calendar/date/CalendarEventViewModel.ts [hunk 17]
```diff
@@ -1281,10 +1350,19 @@ function areRepeatRulesEqual(r1: CalendarRepeatRule | null, r2: CalendarRepeatRu
...
+			areExcludedDatesEqual(r1?.excludedDates ?? [], r2?.excludedDates ?? []))
 	)
 }
 
+/**
+ * compare two lists of dates that are sorted from earliest to latest. return true if they are equivalent.
+ */
+export function areExcludedDatesEqual(e1: ReadonlyArray<DateWrapper>, e2: ReadonlyArray<DateWrapper>): boolean {
+	if (e1.length !== e2.length) return false
+	return e1.find(({ date }, i) => e2[i].date.getTime() !== date.getTime()) == null
```
[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/date/CalendarEventViewModel.ts:1366]: fixed

File: src/calendar/date/CalendarUtils.ts [hunk 18]
```diff
@@ -801,6 +824,55 @@ export function addDaysForRecurringEvent(events: Map<number, Array<CalendarEvent
...
+			if (notExcludedOccurrences.length > 1) {
+				return false
+			}
+			const startTimestamp = startTime.getTime()
+			while (i < excludedTimestamps.length && startTimestamp > excludedTimestamps[i]) {
+				i++
+			}
+
+			if (startTimestamp !== excludedTimestamps[i]) {
+				notExcludedOccurrences.push(startTimestamp)
```
[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/date/CalendarUtils.ts:858]: not quite, but actually storing the occurrences in an array is overkill

File: src/calendar/date/CalendarUtils.ts [hunk 19]
```diff
@@ -801,6 +824,55 @@ export function addDaysForRecurringEvent(events: Map<number, Array<CalendarEvent
...
+	}
+}
+
+/**
+ * find out if a given date is in a list of excluded dates
+ * @param currentDate the date to check
+ * @param excludedDates a sorted list of excluded dates, earliest to latest
+ */
+function isExcludedDate(currentDate: Date, excludedDates: ReadonlyArray<DateWrapper> = []): boolean {
+	return excludedDates.find((dw) => dw.date.getTime() === currentDate.getTime()) != null
```
[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/date/CalendarUtils.ts:873]: fixed

File: src/calendar/export/CalendarImporter.ts [hunk 20]
```diff
@@ -117,7 +121,25 @@ export function serializeRepeatRule(repeatRule: RepeatRule | null, isAllDayEvent
 			endType = `;UNTIL=${value}`
 		}
 
-		return [`RRULE:FREQ=${repeatPeriodToIcalFrequency(assertEnumValue(RepeatPeriod, repeatRule.frequency))}` + `;INTERVAL=${repeatRule.interval}` + endType]
+		const excludedDates = serializeExcludedDates(repeatRule.excludedDates, repeatRule.timeZone)
+		return [
+			`RRULE:FREQ=${repeatPeriodToIcalFrequency(assertEnumValue(RepeatPeriod, repeatRule.frequency))}` + `;INTERVAL=${repeatRule.interval}` + endType,
+		].concat(excludedDates)
```
[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/export/CalendarImporter.ts:127]: the other concats would get lonely

File: src/calendar/export/CalendarParser.ts [hunk 21]
```diff
@@ -353,6 +366,19 @@ export function parseRrule(rruleProp: Property, tzId: string | null): RepeatRule
 	return repeatRule
 }
 
+export function parseExDates(excludedDatesProps: Property[]): DateWrapper[] {
+	const allExDates: Map<number, DateWrapper> = new Map<number, DateWrapper>()
+	for (let excludedDatesProp of excludedDatesProps) {
+		const tzId = getTzId(excludedDatesProp)
+		const values = separatedByCommaParser(new StringIterator(excludedDatesProp.value))
```
[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/export/CalendarParser.ts:373]: this was an exercise and fun

File: src/calendar/view/CalendarEventEditDialog.ts [hunk 22]
```diff
@@ -581,6 +506,110 @@ export async function showCalendarEventDialog(
...
 }
 
+function renderRepeatRulePicker(viewModel: CalendarEventViewModel, startOfTheWeekOffset: number): Children {
+	const intervalValues = createIntervalValues()
+	const endTypeValues = createRepeatRuleEndTypeValues()
+	return [
+		renderTwoColumnsIfFits(
+			[
+				// Repeat type == Frequency: Never, daily, annually etc
+				m(".flex-grow.pr-s", renderRepeatPeriod(viewModel)), // Repeat interval: every day, every second day etc
```
[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/view/CalendarEventEditDialog.ts:516]: fixed

File: src/calendar/view/CalendarEventEditDialog.ts [hunk 23]
```diff
@@ -581,6 +506,110 @@ export async function showCalendarEventDialog(
...
+function renderRepeatPeriod(viewModel: CalendarEventViewModel) {
+	const repeatValues: SelectorItemList<RepeatPeriod | null> = createRepeatRuleFrequencyValues()
+	return m(DropDownSelector, {
+		label: "calendarRepeating_label",
+		items: repeatValues,
+		selectedValue: (viewModel.repeat && viewModel.repeat.frequency) || null,
+		selectionChangedHandler: (period) => viewModel.onRepeatPeriodSelected(period),
+		icon: BootIcons.Expand,
+		disabled: viewModel.isReadOnlyEvent(),
+	} as DropDownSelectorAttrs<RepeatPeriod | null>)
```
[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/view/CalendarEventEditDialog.ts:559]: indeed

[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/view/CalendarEventEditDialog.ts:559]: turns out that at least tests don't compile with this because one of our transformations can't deal with this yet. postponing it until we get to updating them (`as` was used here before anyway)

File: src/calendar/date/CalendarEventViewModel.ts [hunk 24]
```diff
@@ -787,6 +810,43 @@ export class CalendarEventViewModel {
 		}
 	}
 
+	/**
+	 * add an exclusion to the repeat rule of this event, which will cause the event instance that starts at the excluded time to not be rendered or
+	 * fire alarms.
+	 *
+	 * the list of exclusions is maintained sorted from earliest to latest.
+	 */
+	async excludeOccurrence(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (nokhub) [reply src/calendar/date/CalendarEventViewModel.ts]: we changed the functions name and added a comment which should further explain

File: app-ios/tutanotaTests/AlarmManagerTest.swift (hunk 14 above)
Comment (ganthern) [reply app-ios/tutanotaTests/AlarmManagerTest.swift]: hopefully fixed xcode and webstorm do not get along

File: app-ios/tutanota/Sources/Alarms/RepeatRule.swift (hunk 13 above)
Comment (ganthern) [reply app-ios/tutanota/Sources/Alarms/RepeatRule.swift]: yes

File: src/calendar/date/CalendarEventViewModel.ts [hunk 25]
```diff
@@ -226,7 +230,9 @@ export class CalendarEventViewModel {
 		})
 	}
 
-	rescheduleEvent(newStartDate: Date) {
+	// reschedule this event by moving the start and end time by delta milliseconds
```
[the remark this answers is not in the document]
Comment (ganthern) [reply src/calendar/date/CalendarEventViewModel.ts:233]: algorithm does not seem to have changed, just the place where we calculate/apply the offset

File: src/calendar/date/CalendarUtils.ts [hunk 26]
```diff
@@ -729,13 +730,60 @@ export function getRepeatEndTime(repeatRule: RepeatRule, isAllDay: boolean, time
...
+	const isLong = isLongEvent(event, timeZone)
+	const allDay = isAllDayEvent(event)
+	const exclusions = allDay
+		? repeatRule.excludedDates.map(({ date }) => createDateWrapper({ date: getAllDayDateForTimezone(date, timeZone) }))
+		: repeatRule.excludedDates
+
+	iterateOccurrencesWhileTrue(event, timeZone, (startTime: Date, endTime: Date) => {
+		if (startTime.getTime() > month.end.getTime()) return false
+
+		const calcDay = getStartOfDay(startTime)
```
[the remark this answers is not in the document]
Comment (ganthern) [reply src/calendar/date/CalendarUtils.ts:745]: not sure what you mean

[the remark this answers is not in the document]
Comment (charlag) [reply src/calendar/date/CalendarUtils.ts:745]: `getStartOfDay` creates a new `Date` instance and we could sidestep that but it's probably fine

File: src/calendar/date/CalendarEventViewModel.ts [hunk 27]
```diff
@@ -1278,10 +1361,19 @@ function areRepeatRulesEqual(r1: CalendarRepeatRule | null, r2: CalendarRepeatRu
...
+			areExcludedDatesEqual(r1?.excludedDates ?? [], r2?.excludedDates ?? []))
 	)
 }
 
+/**
+ * compare two lists of dates that are sorted from earliest to latest. return true if they are equivalent.
+ */
+export function areExcludedDatesEqual(e1: ReadonlyArray<DateWrapper>, e2: ReadonlyArray<DateWrapper>): boolean {
+	if (e1.length !== e2.length) return false
+	return !e1.some(({ date }, i) => e2[i].date.getTime() !== date.getTime())
```
Comment (charlag) [inline src/calendar/date/CalendarEventViewModel.ts:1374]: would probably easier to read ass `.every()` haha but it's fine

File: src/calendar/view/CalendarViewModel.ts [hunk 28]
```diff
@@ -58,7 +58,9 @@ export type EventsOnDays = {
 	longEvents: Array<CalendarEvent>
 }
 export type DraggedEvent = {
+	/// the event instance the user grabbed with the mouse
```
Comment (charlag) [inline src/calendar/view/CalendarViewModel.ts:61]: this is not jsdoc, `/** */` is jsdoc

File: src/calendar/date/CalendarUtils.ts [hunk 29]
```diff
@@ -729,13 +730,60 @@ export function getRepeatEndTime(repeatRule: RepeatRule, isAllDay: boolean, time
...
+		if (isExcludedDate(startTime, exclusions)) {
+			const eventsOnDay = getFromMap(events, calcDay.getTime(), () => [])
+			findAllAndRemove(eventsOnDay, (e) => isSameEvent(e, event))
```
[the remark this answers is not in the document]
Comment (charlag) [reply src/calendar/date/CalendarUtils.ts:748]: I am still concerned about this?

**Review by charlag — APPROVED** [review]
